### ISSUE TITLE
User switching breaks

### ISSUE THREAD

**Issue by aleixpol** [sddm/sddm#1660]
Split into a separate issue, since it was commented in an already merged PR.

Not sure what you guys are fixing, but I suppose it's not my case.
If I login to user A, then choose "Switch User" from Plasma 5's lockscreen, I get to the SDDM session just fine.
Then if I login to user B, then logout, I get a black screen with a flashing cursor. I can switch the TTY manually, but this just doesn't work for me for years.

Using Wayland with Plasma 5 and sddm-git.

_Originally posted by @smirkybg in https://github.com/sddm/sddm/issues/1644#issuecomment-1418574907_

**Comment by hexchain** [comment]
This issue has been bothering me for a long time. 

Now that logging fixes are in place, I get these messages when logging out of one of the sessions:
```
sddm[837]: Auth: sddm-helper exited successfully
sddm[837]: Socket server stopping...
sddm[837]: Socket server stopped.
sddm[837]: Display server stopping...
sddm[837]: Display server stopped.
sddm[837]: Running display stop script  ("/usr/share/sddm/scripts/Xstop")
sddm[837]: Removing display SDDM::Display(0x558a3ae36a70) ...
sddm[837]: Jumping to VT 0
sddm[837]: VT mode didn't need to be fixed
sddm[837]: Couldn't initiate jump to VT 0: No such device or address
```

Both sessions are Plasma Wayland 5.26.90 and sddm is 0.19.0.184.g58d5c256-2.

**Comment by davidedmundson** [comment]
Are you able to confirm if master of sddm makes a difference?

**Comment by hexchain** [comment]
No, it doesn't. With eb4a7d564cd9385d089ec9ff4ce68042d31ee93c, after logging out of user B it goes to a blank tty with an underline cursor in the top left corner. Logs are the same.

**Comment by BBaoVanC** [comment]
I think I'm getting this issue, looking at `journalctl` of the previous boot (I had to reboot to get out of it), I see a crash loop (about 1-2 times per second):

```
Feb 08 16:01:28 bbaopc sddm[1370]: Adding new display...
Feb 08 16:01:28 bbaopc sddm[1370]: Loaded empty theme configuration
Feb 08 16:01:28 bbaopc sddm[1370]: Xauthority path: "/var/run/sddm/8f1d44d2-c92c-4b61-94a7-57d98e32ccd7"
Feb 08 16:01:28 bbaopc sddm[1370]: Using VT 3
Feb 08 16:01:28 bbaopc sddm[1370]: Display server starting...
Feb 08 16:01:28 bbaopc sddm[1370]: Adding cookie to "/var/run/sddm/8f1d44d2-c92c-4b61-94a7-57d98e32ccd7"
Feb 08 16:01:28 bbaopc sddm[1370]: Running: /usr/bin/X -nolisten tcp -background none -seat seat0 vt3 -auth /var/run/sddm/8f1d44d2-c92c-4b61-94a7-57d98e32ccd7 -noreset -displayfd 16
Feb 08 16:01:29 bbaopc sddm[1370]: Setting default cursor
Feb 08 16:01:29 bbaopc sddm[1370]: Running display setup script  "/usr/share/sddm/scripts/Xsetup"
Feb 08 16:01:29 bbaopc sddm[1370]: Display server started.
Feb 08 16:01:29 bbaopc sddm[1370]: Socket server starting...
Feb 08 16:01:29 bbaopc sddm[1370]: Socket server started.
Feb 08 16:01:29 bbaopc sddm[1370]: Loading theme configuration from "/usr/share/sddm/themes/materia/theme.conf"
Feb 08 16:01:29 bbaopc sddm[1370]: Greeter starting...
Feb 08 16:01:29 bbaopc sddm-helper[25992]: [PAM] Starting...
Feb 08 16:01:29 bbaopc audit[25992]: USER_AUTH pid=25992 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:authentication grantors=pam_permit acct="sddm" exe="/usr/lib/sddm/sddm-helper" hostname=? addr=? terminal=? res=success'
Feb 08 16:01:29 bbaopc audit[25992]: USER_ACCT pid=25992 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:accounting grantors=pam_permit acct="sddm" exe="/usr/lib/sddm/sddm-helper" hostname=? addr=? terminal=? res=success'
Feb 08 16:01:29 bbaopc sddm-helper[25992]: [PAM] Authenticating...
Feb 08 16:01:29 bbaopc sddm-helper[25992]: [PAM] returning.
Feb 08 16:01:29 bbaopc audit[25992]: CRED_ACQ pid=25992 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:setcred grantors=pam_permit acct="sddm" exe="/usr/lib/sddm/sddm-helper" hostname=? addr=? terminal=? res=success'
Feb 08 16:01:29 bbaopc sddm-helper[25992]: pam_unix(sddm-greeter:session): session opened for user sddm(uid=976) by (uid=0)
Feb 08 16:01:29 bbaopc systemd-logind[1309]: New session c39 of user sddm.
░░ Subject: A new session c39 has been created for user sddm
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ Documentation: sd-login(3)
░░ 
░░ A new session with the ID c39 has been created for the user sddm.
░░ 
░░ The leading process of the session is 25992.
Feb 08 16:01:29 bbaopc systemd[1]: Started Session c39 of User sddm.
░░ Subject: A start job for unit session-c39.scope has finished successfully
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit session-c39.scope has finished successfully.
░░ 
░░ The job identifier is 5543.
Feb 08 16:01:29 bbaopc audit[25992]: USER_START pid=25992 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:session_open grantors=pam_unix,pam_systemd acct="sddm" exe="/usr/lib/sddm/sddm-helper" hostname=? addr=? terminal=:0 res=success'
Feb 08 16:01:29 bbaopc sddm-helper[25992]: Can't start X11 session with empty auth cookie
Feb 08 16:01:29 bbaopc sddm[1370]: Greeter session failed to start
Feb 08 16:01:29 bbaopc audit[25992]: USER_END pid=25992 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:session_close grantors=pam_unix,pam_systemd acct="sddm" exe="/usr/lib/sddm/sddm-helper" hostname=? addr=? terminal=:0 res=success'
Feb 08 16:01:29 bbaopc audit[25992]: CRED_DISP pid=25992 uid=0 auid=4294967295 ses=4294967295 msg='op=PAM:setcred grantors=pam_permit acct="sddm" exe="/usr/lib/sddm/sddm-helper" hostname=? addr=? terminal=:0 res=success'
Feb 08 16:01:29 bbaopc sddm-helper[25992]: [PAM] Closing session
Feb 08 16:01:29 bbaopc sddm-helper[25992]: pam_unix(sddm-greeter:session): session closed for user sddm
Feb 08 16:01:29 bbaopc sddm-helper[25992]: [PAM] Ended.
Feb 08 16:01:29 bbaopc sddm[1370]: Auth: sddm-helper exited with 2
Feb 08 16:01:29 bbaopc sddm[1370]: Greeter stopped. SDDM::Auth::HELPER_SESSION_ERROR
Feb 08 16:01:29 bbaopc sddm[1370]: Socket server stopping...
Feb 08 16:01:29 bbaopc sddm[1370]: Socket server stopped.
Feb 08 16:01:29 bbaopc sddm[1370]: Display server stopping...
Feb 08 16:01:29 bbaopc systemd-logind[1309]: Session c39 logged out. Waiting for processes to exit.
Feb 08 16:01:29 bbaopc systemd[1]: session-c39.scope: Deactivated successfully.
░░ Subject: Unit succeeded
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ The unit session-c39.scope has successfully entered the 'dead' state.
Feb 08 16:01:29 bbaopc systemd-logind[1309]: Removed session c39.
░░ Subject: Session c39 has been terminated
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ Documentation: sd-login(3)
░░ 
░░ A session with the ID c39 has been terminated.
Feb 08 16:01:29 bbaopc sddm[1370]: Display server stopped.
Feb 08 16:01:29 bbaopc sddm[1370]: Running display stop script  ("/usr/share/sddm/scripts/Xstop")
Feb 08 16:01:29 bbaopc sddm[1370]: Removing display SDDM::Display(0x56535a3670d0) ...
```

**Comment by hexchain** [comment]
@BBaoVanC Did you just update SDDM from a version before the Xauth handling redesign (cb5c07b4c5bafa27b21981a8f4039eadbe0fafdf, eb4a7d564cd9385d089ec9ff4ce68042d31ee93c) to something after? I had the same problem after such an upgrade but IMO it isn't quite related to this issue. The crash loop should not happen again once rebooted.

**Comment by wen-zheng** [comment]
encountered same problem

**Comment by BBaoVanC** [comment]
> @BBaoVanC Did you just update SDDM from a version before the Xauth handling redesign ([cb5c07b](https://github.com/sddm/sddm/commit/cb5c07b4c5bafa27b21981a8f4039eadbe0fafdf), [eb4a7d5](https://github.com/sddm/sddm/commit/eb4a7d564cd9385d089ec9ff4ce68042d31ee93c)) to something after? I had the same problem after such an upgrade but IMO it isn't quite related to this issue. The crash loop should not happen again once rebooted.

That must have been it, there's no more crashing now.

**Comment by Vogtinator** [comment]
Hm, indeed, #1230 changed the protocol between sddm and sddm-helper such that a running (old) sddm won't be able to start an X11 session with the new sddm-helper. Not sure what to do about that, maybe sddm should restart itself if it detects that it was updated.

**Comment by Vogtinator** [comment]
> Hm, indeed, https://github.com/sddm/sddm/pull/1230 changed the protocol between sddm and sddm-helper such that a running (old) sddm won't be able to start an X11 session with the new sddm-helper.

There are two differences:

The greeter session previously used `XAUTHORITY=/run/sddm/...` while now it's started with a cookie just like a user session. The previous behaviour of sddm trips the `Can't start X11 session with empty auth cookie` check in `sddm-helper`.

The cookie is passed as `QByteArray` instead of `QString`.

**Comment by rapgro** [comment]
maybe #1520 is related as regression

**Comment by MrAntonS** [comment]
Encountered same issue

**Comment by wen-zheng** [comment]
encountered same issue

**Comment by rottenpants466** [comment]
Same problem here.

**Comment by Vogtinator** [comment]
Fixed with #1690
